src/solvers/dgsem/basis_lobatto_legendre.jl

```diff
@@ -196,47 +214,79 @@ function Base.show(io::IO, ::MIME"text/plain", mortar::LobattoLegendreMortarL2)
           polydeg(mortar))
 end
 
-@inline Base.real(mortar::LobattoLegendreMortarL2{RealT}) where {RealT} = RealT
+abstract type AbstractMortarEC{RealT} <: AbstractMortar{RealT} end
+
+struct LobattoLegendreMortarEC{RealT <: Real, NNODES,
+                               ForwardMatrix <: AbstractMatrix{RealT},
+                               ReverseMatrix <: AbstractMatrix{RealT}} <:
+       AbstractMortarEC{RealT}
+    forward_upper::ForwardMatrix
+    forward_lower::ForwardMatrix
+    reverse_upper::ReverseMatrix
+    reverse_lower::ReverseMatrix
+end
+
+"""
+    MortarEC(basis::LobattoLegendreBasis)
+
+!!! warning "Experimental implementation"
+    This is an experimental feature and may change in any future releases.
+    Currently only implemented for 2D [`TreeMesh`](@ref).
+
+Flux-corrected, entropy-conservative L2-projection based mortar tailored to the [`LobattoLegendreBasis`](@ref).
+Mortars are required to handle non-conforming interfaces. 
+Currently, Trixi.jl only supports 2h-nonconforming interfaces.
+This means that all nonconforming interfaces (lines in 2D and faces in 3D) are of a 2-1 (or 1-2) coupling, i.e.,
+side A has exactly twice the number of nodes (per dimension) of side B (or vice-versa).
+
+The entropy-conservative mortar is based on the [`MortarL2`](@ref) but adds a flux correction term
+to ensure that the mortar projection/interpolation operations are entropy-conservative.
+
+# References
+- Jesse Chan, Mario J. Bencomo, and David C. Del Rey Fernández (2021).
+  Mortar-based Entropy-Stable Discontinuous Galerkin Methods on Non-conforming Quadrilateral and Hexahedral Meshes.
+  [DOI:10.1007/s10915-021-01652-3](https://doi.org/10.1007/s10915-021-01652-3)
+"""
+function MortarEC(basis::LobattoLegendreBasis)
+    RealT = real(basis)
+    nnodes_ = nnodes(basis)
+
+    forward_upper = calc_forward_upper(nnodes_, RealT)
+    forward_lower = calc_forward_lower(nnodes_, RealT)
+    reverse_upper = calc_reverse_upper(nnodes_, Val(:gauss), RealT)
+    reverse_lower = calc_reverse_lower(nnodes_, Val(:gauss), RealT)
+
+    LobattoLegendreMortarEC{RealT, nnodes_, typeof(forward_upper),
+                            typeof(reverse_upper)}(forward_upper, forward_lower,
+                                                   reverse_upper, reverse_lower)
+end
+
+function Base.show(io::IO, mortar::LobattoLegendreMortarEC)
+    @nospecialize mortar # reduce precompilation time
+
+    print(io, "LobattoLegendreMortarEC{", real(mortar), "}(polydeg=", polydeg(mortar),
+          ")")
+end
+function Base.show(io::IO, ::MIME"text/plain", mortar::LobattoLegendreMortarEC)
+    @nospecialize mortar # reduce precompilation time
+
+    print(io, "LobattoLegendreMortarEC{", real(mortar), "} with polynomials of degree ",
+          polydeg(mortar))
+end
+
+@inline Base.real(mortar::Union{LobattoLegendreMortarL2{RealT},
+LobattoLegendreMortarEC{RealT}}) where {RealT} = RealT
```

Comment:
This is quite ugly formatting :(

Reply:
```suggestion
@inline function Base.real(mortar::Union{LobattoLegendreMortarL2{RealT},
                                         LobattoLegendreMortarEC{RealT}}) where {RealT}
    return RealT
end
```
?

Reply:
Can we maybe define

```julia
const LobattoLegendreMortar{RealT, NNODES} = Union{LobattoLegendreMortarL2{RealT, NNODES},
                                                   LobattoLegendreMortarEC{RealT, NNODES}}
```

to make dispatch nicer?